File: programmers/level2/[PCCP 기출문제] 2번 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    answer = 0
    start = 1  # 레벨 1부터
    end = 10 ** 15  # limit의 크기까지

    def caculate(level):
        sum = 0  # 소요시간 계산
        for i in range(len(diffs)):
            if diffs[i] <= level:  # 퍼즐 틀리지 않는 경우
                sum += times[i]  # 퍼즐 해결한 경우
            else:
                tc = times[i]  # 현재 시간
                tp = 0  # 이전 시간
                if i - 1 >= 0:  # 첫번째 원소가 아닌 경우
                    tp = times[i - 1]  # 이전 시간 셋팅
                sum += (tc + tp) * (diffs[i] - level) + tc  # 소요시간 셋팅
        return sum

    while True:
        if start >= end:  # 탐색 종료 조건
            break

        mid = (end + start) // 2  # 중간값 계산
        sum = caculate(mid)  # 중간값을 레벨로 체크

        if sum > limit:  # 합계가 더 크다면 렙이 더 올라가야함
            start = mid + 1
        else:  # 합계가 더 작다면 렙을 낮춰야함
            end = mid

    answer = start  # 가장 최솟값이 렙
    return answer

File: programmers/level2/[PCCP 기출문제] 2번 퍼즐 게임 챌린지.java
import java.util.*;
import java.lang.Math;

class Solution {
    public long check(int[] diffs, int[] times, long level) {
        long time = 0;

        for (int i = 0; i < diffs.length; i++) {
            if (diffs[i] <= level) {
                time += times[i];
            } else {
                int nowTime = times[i];
                int prevTime = i == 0 ? 0 : times[i-1];
                time += (nowTime + prevTime) * (diffs[i] - level) + nowTime;
            }
        }
        return time;
    }

    public long solution(int[] diffs, int[] times, long limit) {
        long answer = 0;
        long start = 0;
        long end = (long) Math.pow(10, 15);
        long mid = 0;
        long sum = 0;

        while (start <= end) {
            mid = (start + end) / 2;
            sum = check(diffs, times, mid);

            // System.out.println("start = " + start + " end = " + end + " mid = " + mid + " sum = "+ sum);

            if (sum <= limit) {
                end = mid - 1;
                answer = mid;
            } else {
                start = mid + 1;
            }
        }

        if (answer == 0) { // 모두 해결이 가능했어서 숙련도 연산이 되지 않았을 경우
            answer = 1;
        }
        return answer;
    }
}